File: LinkedList/No24.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode swapPairs(ListNode head) {
        ListNode dummy = new ListNode(Integer.MIN_VALUE);
        dummy.next = head;
        
        ListNode pre = dummy;
        ListNode a = null;
        ListNode b = null;
        
        while(pre.next != null && pre.next.next != null){
            //to swap
            a = pre.next;
            b = a.next;
            //swap
            pre.next = b;
            a.next = b.next;
            b.next = a;
            //state change
            pre = pre.next.next;
        }
        
        return dummy.next;
    }
}

File: mid/No24.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def swapPairs(self, head):
        
        def recur(node):
            if not node:
                return None
            rt = node.next
            if not rt:
                return node
            node.next = rt.next
            rt.next = node
            node.next = recur(node.next)
            return rt
        
        dummy = ListNode(-1)
        dummy.next = head
        dummy.next = recur(dummy.next)
        return dummy.next
    
        
        
